docker-compose.yml

```diff
@@ -0,0 +1,87 @@
+version: "3"
+
+services:
+  database:
+    container_name: DarkflameDatabase
+    image: mariadb:10.6
+    environment:
+      - MARIADB_USER=${MARIADB_USER:-darkflame}
+      - MARIADB_PASSWORD=${MARIADB_PASSWORD:-darkflame}
+      - MARIADB_ROOT_PASSWORD=${MARIADB_ROOT_PASSWORD:-darkflame}
+      - MARIADB_DATABASE=${MARIADB_DATABASE:-darkflame}
+    volumes:
+      - database:/var/lib/mysql
+      - ./migrations/dlu:/docker-entrypoint-initdb.d
```

Comment:
How does the migrations wind up working if you run an old version of the container, then the source gets updated with new migrations, then you update the container to use the latest version?

Reply:
As far as I understand it, this only runs if the database is unintialized (ie. the database volume is empty), so the migration SQL never runs unless you wipe all your data.

I've been running it this way for the past couple of days and many container restarts without any issues.

Reply:
From the docs here: https://hub.docker.com/_/mariadb

> ### Initializing a fresh instance
> When a container is started for the first time, a new database with the specified name will be created and initialized with the provided configuration variables. Furthermore, it will execute files with extensions .sh, .sql, .sql.gz, .sql.xz and .sql.zst that are found in /docker-entrypoint-initdb.d. Files will be executed in alphabetical order. .sh files without file execute permission are sourced rather than executed. You can easily populate your mariadb services by mounting a SQL dump into that directory and provide custom images with contributed data. SQL files will be imported by default to the database specified by the MARIADB_DATABASE / MYSQL_DATABASE variable.

Reply:
Ideally any un-applied migrations could be run automatically (which may mean creating a migrations table or such) - or at least document a clear command to run a specific migration file so that its clear how it can be done manually

Reply:
Is this safe? Could the database reach a state which lets new migrations fail, because the DB has been changed by the server? (Not sure if this is possible, but if it is this should be kept in mind for new migrations.)
And if you want to create this table in case of "merge conflicts" (I may have misunderstood), it may be too much to ask for an inexperienced user to merge that themselves.

Reply:
This would also require the tables that are changed by the migrations not being changed by the user (or the server), because otherwise new migration files could fail, couldn't they? If they are only created with new servers in mind?

Reply:
The structure of the DB should only ever be changed by migrations (and everything after the initial migrations should preserve any existing data in a DB) - except for if you're doing development and writing WIP migrations, in which case all bets are off (and frankly, in that case I don't think we need to worry about data preservation and such, right?)

Reply:
Well in that case this is no issue.

Reply:
Yeah I think the key here is that loading the file in this way is safe since it will only be run at database init. Any other migrations are up to the user and should be done manually. This is just the initialization file and will only run once.

Reply:
I think what @luxaritas  is saying is that we can freely apply the migrations, even after the db is initialized, because the database scheme is not changed by the server or the user.
The point would be that we can apply new migrations anytime, but have to keep track of the ones we already applied, so we don't apply them again.